File: IdeaProjects/ScalaTutorial/ScalaMessageReader/src/main/scala/scalaMessageReader/endpoint/OutputToCSVExporter.scala
package scalaMessageReader.endpoint
import java.io.{BufferedWriter, FileWriter}

import org.springframework.messaging.Message

import scala.collection.mutable
import scalaMessageReader.Notification
import scalaMessageReader.MessageReaderConstants._

class OutputToCSVExporter extends MessageEndpoint {

  final val DELIMETER = ","

  override def handleMessage(message: Message[Notification]) : Unit = {
    val messageMap : mutable.Map[String,String] = message.getPayload.messageDetails
    messageMap.put(END_TIME, message.getPayload.endProcessingTime)
    if (messageMap.contains(START_TIME)) exportToCSV(messageMap)
  }

  def exportToCSV(message : mutable.Map[String,String]) : Unit = {
    val fileWriter = new FileWriter(RESULTS_FILE_PATH + RESULTS_FILE_NAME, true)
    val bw = new BufferedWriter(fileWriter)
    bw.append(message(ENTITY_NAME)).append(DELIMETER)
          .append(message(START_TIME)).append(DELIMETER)
          .append(message(END_CONVERSION_TIME)).append(DELIMETER)
          .append(message(END_TIME)).append(DELIMETER)
          .append(message(STATUS))
    bw.newLine()
    bw.close()
    fileWriter.close()
  }
}

File: IdeaProjects/MessageReader/src/main/java/messageReader/endpoint/OutputToCSVExporter.java
package messageReader.endpoint;

import messageReader.Notification;
import org.springframework.messaging.Message;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Map;

import static messageReader.MessageReaderConstants.*;


public class OutputToCSVExporter implements MessageEndpoint {

    public OutputToCSVExporter() throws IOException {
    }

    @Override
    public void handleMessage(Message<Notification> message) throws IOException {
        Map<String, String> messageMap = message.getPayload().getMessageDetails();
        messageMap.put(END_TIME, message.getPayload().getEndProcessingTime());
        if (hasStartAndEndTimes(messageMap)) {
            exportToCsv(messageMap);
        }
    }

    private void exportToCsv(Map<String, String> message) throws IOException {
        FileWriter fileWriter = new FileWriter(RESULTS_FILE_PATH + RESULTS_FILE_NAME, true);
        BufferedWriter bw = new BufferedWriter(fileWriter);

        bw.append(message.get(ENTITY_NAME)).append(DELIMETER)
                .append(message.get(START_TIME)).append(DELIMETER)
                .append(message.get(END_CONVERSION_TIME)).append(DELIMETER)
                .append(message.get(END_TIME)).append(DELIMETER)
                .append(message.get(STATUS));
        bw.newLine();

        bw.close();
        fileWriter.close();
    }

    private boolean hasStartAndEndTimes(Map<String, String> message) {
        return message.containsKey(START_TIME) && message.containsKey(END_TIME);
    }


}
